### DISCUSSION TITLE
Hi Rector Team. What might be a possible reason that Rector stops processing at 26%?

### DISCUSSION THREAD

**Discussion by alexandreelise** [rectorphp/rector#7403] [Q&A] [answered]
Hi,

Since I can't find why latest rector 0.14 won't work with my setup. I downgrade it to rector ^0.12
There is no more fatal error of class redifinition of phpstan dependency but for some reason, rector process the files stops at 26% and say Rector done. Any hint @TomasVotruba ?

**Comment by TomasVotruba** [comment, marked as answer]
Hi, I'd try narrow scoping files you provide.
Also `--debug` option might give some helpeful info about which file/rule stops it on.

If you want exact advice, we'll need a reproducible Github repository to be able to tell for sure :)

**Reply by alexandreelise** [reply]
Hi, tomas,
Even with the --debug flag rector keeps telling me Rector is done without verbose output even when using it with --dry-run.

**Reply by TomasVotruba** [reply]
I see.

We'll need a reproducible Github repository to be able to tell for sure :)

**Reply by alexandreelise** [reply]
Unfortunately. This particular project is a private project and we cannot share the code.

**Reply by TomasVotruba** [reply]
Not neccessary to cope whole project :) usually it's a file or two, that cause the error. 
E.g. this is size of good-enough reproducer: https://getrector.org/demo

**Reply by alexandreelise** [reply]
Hi,
It seems to be a chiken-egg problem. As we use rector to downgrade code to project's minimum php but can't do it since we cannot find the origin of the issue which prevents us of downgrading and so on.
Nevertheless thanks for your piece of advise we'll dig in and try harder. Rector is an awesome tool and we want to use it to the maximum it can do. We'll get there. Thanks again Tomas and have a great day rectifying legacy code automatically.

**Reply by TomasVotruba** [reply]
Thank you for sincere words :bow: 

I wish you find your solution and make it work :+1:

**Reply by alexandreelise** [reply]
Hey Tomas!
There is some substantial progress. We red the whole documentation to grasp rector better. We used ecs to fix messy code. we passed everything through grumphp which has some linters on pre-commit hook. We fixed fatal errors. Now rector is happy and show the debug and analyse the files fully. But one last thing we couldn't make rector downgrade the code to older php version the composer platform is 7.2.5 composer require php is ^7.2.5  but when we configure rector with 

```
// define sets of rules
	$rectorConfig->sets([
		LevelSetList::UP_TO_PHP_72,
		SetList::PHP_72,
		SetList::DEAD_CODE,
		SetList::MYSQL_TO_MYSQLI,
		SetList::NAMING,
	]);

```

and

```
// PHP minimum for Joomla! 4.x
	$rectorConfig->phpVersion(70205);

```

It doesn't downgrade arrow function for example.

Can you please shed a light on that one?

We are almost there we need help from Rector Creator. Thanks again Tomas.

**Reply by samsonasik** [reply]
To downgrade, you need to use rector-downgrade-php DowngradeLevelSetList https://github.com/rectorphp/rector-downgrade-php#use-sets, see 

https://github.com/rectorphp/rector-src/blob/6df313efe5709797e53fe50275da97341ed79295/build/config/config-downgrade.php#L20

for example

**Reply by alexandreelise** [reply]
Hi @samsonasik What is the difference between UP_TO_PHP_72 and DOWN_TO_PHP_72?
When I try to use DOWN_TO_PHP_72 rector throws TypeError at 26/93 files to process whereas with UP_TO_PHP_72 it continues until the end. Can you shed a light why? Thanks again.

**Reply by alexandreelise** [reply]
Hi @samsonasik It downgraded the code as expected but only when I remove the SetList::DEADCODE. Apparently as of Rector 0.12.23  , rector doesn't do downgrade and remove dead code espescially remove unsused private method at the same time. Just to give you my feedback. I might have been fixed in the latest version of rector the 0.14.0

**Reply by samsonasik** [reply]
when downgrading, I suggest to not include other sets, when it succesfully downgraded, then after it, you can apply other setlist.

**Reply by alexandreelise** [reply]
Thanks for your help @samsonasik

**Reply by alexandreelise** [reply]
Hi Rector team. You can mark this one as solved.
